python/pyarrow/parquet.py

```diff
@@ -1252,6 +1341,95 @@ def _make_manifest(path_or_paths, fs, pathsep='/', metadata_nthreads=1,
     return pieces, partitions, common_metadata_path, metadata_path
 
 
+class _ParquetDatasetV2:
+    """
+    ParquetDataset shim using the Dataset API under the hood.
+    """
+    def __init__(self, path_or_paths, filesystem=None, filters=None,
+                 read_dictionary=None, buffer_size=None, **kwargs):
+        import pyarrow.dataset as ds
+        import pyarrow.fs
+
+        # Raise error for not supported keywords
+        for keyword, default in [
+                ("schema", None), ("metadata", None),
+                ("split_row_groups", False), ("validate_schema", True),
+                ("metadata_nthreads", 1), ("memory_map", False)]:
+            if keyword in kwargs and kwargs[keyword] is not default:
+                raise ValueError(
+                    "Keyword '{0}' is not yet supported with the new "
+                    "Dataset API".format(keyword))
+
+        # map old filesystems to new one
+        # TODO(dataset) deal with other file systems
+        if isinstance(filesystem, LocalFileSystem):
+            filesystem = pyarrow.fs.LocalFileSystem()
```

Comment:
In `pyarrow.dataset` memory mapping is deferred to the filesystem which opens the files, so `memory_map` should probably be supported with:
```suggestion
        if isinstance(filesystem, LocalFileSystem):
            filesystem = pyarrow.fs.LocalFileSystem(use_mmap=memory_map)
```

Reply:
Should we always enable memory mapping? Eg will it be beneficial for something like Parquet that needs to be decompressed anyway?

Reply:
Actually, I see that we right now have a `memory_map` that I coded as unsupported, but will let it trigger this. The question is what to do if the LocalFileSystem is already passed as an object ..